istioctl/cmd/webhook.go

```diff
@@ -0,0 +1,714 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/ghodss/yaml"
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/spf13/cobra"
+
+	"k8s.io/api/admissionregistration/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	caCertServiceAccount          = "default"
+	caCertServiceAccountNamespace = "default"
+	caKeyInK8sSecret              = "ca.crt"
+	certKeyInK8sSecret            = "cert-chain.pem"
+)
+
+type enableCliOptions struct {
+	// Whether enable the webhook configuration of Galley
+	enableValidationWebhook bool
+	// Whether enable the webhook configuration of Sidecar Injector
+	enableMutationWebhook bool
+	// The local file path of webhook CA certificate
+	caCertPath string
+	// The file path of the validation webhook configuration.
+	validationWebhookConfigPath string
+	// The service name of the validating webhook to manage
+	validatingWebhookServiceName string
+	// The file path of the mutating webhook configuration.
+	mutatingWebhookConfigPath string
+	// The service name of the mutating webhook to manage
+	mutatingWebhookServiceName string
+	// Max time for checking the validating webhook.
+	// If the validating webhook is not ready in the given time, exit.
+	// Otherwise, apply the webhook configuration.
+	maxTimeForCheckingWebhookServer time.Duration
+	// Max time for waiting the webhook certificate to be readable.
+	maxTimeForReadingWebhookCert time.Duration
+	// The name of the secret of a webhook.
+	// istioctl will check whether the certificate in the secret is issued by
+	// the certificate in the Kubernetes service account or the given CA cert, if any.
+	webhookSecretName string
+	// The namespace of the webhook secret.
+	webhookSecretNameSpace string
+}
+
+// Validate return nil if no error. Otherwise, return the error.
+func (opts *enableCliOptions) Validate() error {
+	if !opts.enableValidationWebhook && !opts.enableMutationWebhook {
+		return fmt.Errorf("no webhook to enable")
+	}
+	if opts.enableValidationWebhook {
+		if len(opts.validationWebhookConfigPath) == 0 {
+			return fmt.Errorf("must specify a valid --validation-path of the validation webhook configuration")
```

Comment:
Maybe `--validation-path <yaml-file> is required`.  We aren't checking that it is valid, that will be checked later when we try to open the file.  Same comment for `--validation-service` and `--webhook-secret`.  These are required, and this is the place that checks that they are present.  The message should give a hint if this is a Kubernetes .YAML of a particular kind, or a Kubernetes object on a live cluster, etc.

Reply:
Done.